add invoice transport parsing

nicbus opened the pull request:
This PR adds parsing for the `RgbTransport` from the invoice's scheme.

The format for the transport I chose is `<transport><tls>-<host>`, where:
- `<transport>` is the transport name (with no `-` character, as it's used to separate the host portion)
- `<tls>` is `0` for cleartext and `1` is for tls
- `<host>` is the transport host (can contain `-` characters)

I (temporarily) commented RestHttp, WebSockets and Storm variants as they're now unused.

I also moved the `map_query_params` function out of `RgbInvoice` as it doesn't use self (as noted in a previous PR).

In the context of issue #34, I kept the tls parameter (to avoid breaking URL integrity) but used 0/1 for brevity and didn't use the `+` charachter, which I plan to use to support multiple transports (did a minimal test implementation and it appears to work as intended)

dr-orlovsky commented:
Sorry, I do not understand the approach: it makes invoices to be invalid URIs, putting host into schema. What is the reason?

zoedberg commented:
> Sorry, I do not understand the approach: it makes invoices to be invalid URIs, putting host into schema. What is the reason?

I think the reason for this is to allow future support for multiple `RgbTransport`s, each having it's host and then joined with a `+`. This way invoice uri scheme part will contain the transports info and rest of uri will contain all other payment info.

I don't think this makes invoices invalid, uri library is parsing the string correctly, but if you have a better proposal to support multiple rgb transports we'll be happy to hear that

dr-orlovsky commented:
Sorry, approach NACK. Adding host to the URI schema makes for instance impossible to register apps for working with RGB. There are a lot of other problems with the approach (inability to specify port number, incompatibility with IPv6 addresses). We need to find a different way of adding multiple transports - I will look more into that.

zoedberg commented:
> There are a lot of other problems with the approach (inability to specify port number, incompatibility with IPv6 addresses).

You're right 

> We need to find a different way of adding multiple transports - I will look more into that.

Maybe we can always have `rgb:` as scheme and add multiple transports info to the query part? So that if transport query field is missing we assume an "unspecified means" transport, and otherwise the invoice can embed one or more transports to the query part

dr-orlovsky commented:
Yes, I came to the same conclusion after contemplating on the question. I think the query is the right place to put it. We may have an `alt` query parameter, which takes a list of alternative transports and hosts. It can look like:

`rgb-rpc://host1/..../?alt=storm://node_key@host2,ws`

The rules can be following:
- an alternative host can be given next to each alternative transport in its normal form. I have checked that [all necessary chars are allowed in the query "URL entity"](https://url.spec.whatwg.org/#url-code-points) (like `:`, `/`, `@`, `.`), which makes it really readable
- if an alternative host is omitted (like in `ws` case in the sample above) than host from the main URL is used
- `rgb` prefix in all alternative transports should be omitted (less visual clutter + we have rgb prefix for the whole URL).

I also propose to stick to the original transport names, which are a combination of `rgb` prefix with either
* existing transport (usable outside of rgb, like `ws`, `storm`, `http`) - in which case a `+` char must be used
* a transport which is defined only in context of RGB, i.e. `rgb-rpc` for now, in which case ` -` char must be used.

dr-orlovsky commented:
There is another viable alternative: always omit transport and host from an RGB URL and provide them (when needed) as a `transport` or `endpoints` query parameter. The above in this case will look this way:

`rgb:contractName/ifaceName?alt=rpc://host1,storm://node_key@host2,ws`

I personally do not yet know which way is better. This one looks more consistent, but it will leave us with just a single URL scheme name independent of transport, limiting user ability to use different wallets each of which support specific transports. On the other side, a user paying invoice must be able to pay it from the wallet having RGB assets and the transport should not play a part in selecting an app to open. So probably option with single RGB scheme and all transports given as query param may be most preferable in all the cases...

dr-orlovsky commented:
The more I think the more I like the second option

cryptoquick commented:
I prefer the second, also. For example, the data could be stored on a Carbonado node's HTTP frontend, and if one node was down, others can be provided as backups. Kind of like Nostr's relays system. I think `alt=` should just be `src=`.

nicbus commented:
I agree that the second proposal is the best one and have reworked transport parsing to match.

Multiple endpoint support has also been introduced, changing the `RgbInvoice` field name to `transports`, which is now a `Vec<RgbTransport>`.

Given the new transport specification is closer to a typical URL, I switched from `0`/`1` to using an additional `s` to indicate TLS support, as IMO it feels more natural.

Replaced the derived `Display` trait for `RgbTransport` with a manual implementation, in order to achieve symmetry between invoice parsing and displaying.

Split `RgbTransport` parsing errors to a dedicated `TransportParseError` enum for consistency.

Replaced `urlencoding` crate with `percent-encoding` to have more control on which characters need to be %-encoded. I derived the query %-encode set from https://url.spec.whatwg.org/#query-percent-encode-set and added:
- the characters `[` and `]`, as url parsing fails when those chars are used
- the characters `&` and `=`, as those have special meaning in query params

Tests have been updated to cover previously unhandled cases, like specifying host authentication and port, using an IPv6 address, more transport types, multiple endpoints.

dr-orlovsky commented on src/invoice.rs:
```diff
@@ -87,6 +146,18 @@ pub struct RgbInvoice {
+}
+
 #[derive(Clone, PartialEq, Eq, Debug, Display, Error, From)]
 #[display(inner)]
```
Nit: we ca do
```suggestion
#[display(doc_comments)]
```

and remove all `doc_comments` in the variants - adding `#[display(inner)]` where they are absent

dr-orlovsky reviewed (approved):
ACK 7eaa76402820329adcca4dde08486446ccf14406
